### PR TITLE
Adapt version backend to new features

### PR DESCRIPTION
**Pull request by artonge**
Adapt the version backend to the new files versions feature.

## New
+ Support naming a version
+ Support deleting a version
+ Adds DB table to store version infos
+ Fix version creation that was using the current time instead of the version mtime
+ Adapt `LazyFolder` to new `FileInfo` interface https://github.com/nextcloud/server/pull/38150

## Todo
- [x] Test version import from previous versions
- [x] Check tests

Fix https://github.com/nextcloud/server/issues/39138
Needs https://github.com/nextcloud/server/pull/40296

### CODE REVIEW

**Review by AndyScherzinger — APPROVED** [review]

File: lib/Migration/Version16000Date20230821085801.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		]);
+		$table->addColumn('timestamp', Types::BIGINT, [
+			'notnull' => true,
+			'length' => 20,
+		]);
+		$table->addColumn('size', Types::BIGINT, [
+			'notnull' => true,
+			'length' => 20,
+		]);
+		$table->addColumn('mimetype', Types::BIGINT, [
```
Comment (come-nc) [inline lib/Migration/Version16000Date20230821085801.php:72]: BIGINT for mimetype? :thinking:

File: lib/Versions/GroupVersion.php [hunk 2]
```diff
@@ -46,12 +41,11 @@ public function __construct(
...
-		File $versionFile,
-		int $folderId
+		string $label,
+		private File $versionFile,
+		private int $folderId,
 	) {
-		parent::__construct($timestamp, $revisionId, $name, $size, $mimetype, $path, $sourceFileInfo, $backend, $user);
-		$this->versionFile = $versionFile;
-		$this->folderId = $folderId;
+		parent::__construct($timestamp, $revisionId, $name, $size, $mimetype, $path, $sourceFileInfo, $backend, $user, $label);
```
Comment (come-nc) [inline lib/Versions/GroupVersion.php:48]: Both min and max version will need to be 28 then? Is that already the case in info.xml?

File: lib/Versions/GroupVersionEntity.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+	public function setLabel(string $label): void {
+		$metadata = $this->getDecodedMetadata();
+		$metadata['label'] = $label;
+		$this->setDecodedMetadata($metadata);
+		$this->markFieldUpdated('metadata');
+	}
+
+	public function getDecodedMetadata(): array {
+		return json_decode($this->metadata ?? '', true) ?? [];
```
Comment (come-nc) [inline lib/Versions/GroupVersionEntity.php:85]: It should throw on error as well, no?

File: lib/Versions/VersionsBackend.php [hunk 4]
```diff
@@ -32,66 +35,118 @@
...
+			// Insert the entry in the DB for the current version.
+			$versionEntity = new GroupVersionEntity();
+			$versionEntity->setFileId($file->getId());
+			$versionEntity->setTimestamp($file->getMTime());
+			$versionEntity->setSize($file->getSize());
+			$versionEntity->setMimetype($this->mimeTypeLoader->getId($file->getMimetype()));
+			$versionEntity->setDecodedMetadata([]);
+			$this->groupVersionsMapper->insert($versionEntity);
+
+			// Insert entries in the DB for existing versions.
```
Comment (come-nc) [inline lib/Versions/VersionsBackend.php:91]: I’m lost, why is that needed outside of migrations?

File: lib/Versions/VersionsBackend.php [hunk 5]
```diff
@@ -32,66 +35,118 @@
...
+				$versionEntity->getTimestamp(),
+				$file->getName(),
+				$versionEntity->getSize(),
+				$this->mimeTypeLoader->getMimetypeById($versionEntity->getMimetype()),
+				$userFolder->getRelativePath($file->getPath()),
+				$file,
+				$this,
+				$user,
+				$versionEntity->getLabel(),
+				$file->getMtime() === $versionEntity->getTimestamp() ? $file : $versionsFolder->get((string)$versionEntity->getTimestamp()),
```
Comment (come-nc) [inline lib/Versions/VersionsBackend.php:143]: Is there any check to avoid a mess if two versions are created within 1 second?

File: lib/Versions/VersionsBackend.php (hunk 4 above)
Comment (come-nc) [reply lib/Versions/VersionsBackend.php:91]: Ok, got it, outside of migration the function called before will find versions in DB and this will not run.

File: lib/Versions/VersionsBackend.php [hunk 6]
```diff
@@ -211,4 +267,67 @@ private function getVersionsFolder(int $folderId): Folder {
...
+		$sourceFile = $version->getSourceFile();
+		$mount = $sourceFile->getMountPoint();
+
+		if (!($mount instanceof GroupMountPoint)) {
+			return;
+		}
+
+		$versionsFolder = $this->getVersionsFolder($mount->getFolderId())->get((string)$sourceFile->getId());
+		/** @var Folder $versionsFolder */
+		$versionsFolder->get((string)$version->getTimestamp())->delete();
```
Comment (come-nc) [inline lib/Versions/VersionsBackend.php:293]:
```suggestion
		$versionsFolder->get((string)$version->getRevisionId())->delete();
```
?
I know it should match the timestamp but…

**Comment by artonge** [comment]
For the backport
- [ ] Bump app version in `info.xml` to trigger the migration

File: lib/Migration/Version16000Date20230821085801.php (hunk 1 above)
Comment (artonge) [reply lib/Migration/Version16000Date20230821085801.php:72]: Because https://github.com/nextcloud/server/blob/6f80ae274bb7795eab9517a4278d394e8da156fb/core/Migrations/Version13000Date20170718121200.php#L164-L168
